phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java

```diff
@@ -46,27 +61,37 @@
 import org.apache.phoenix.parse.HintNode.Hint;
 import org.apache.phoenix.query.KeyRange;
 import org.apache.phoenix.query.KeyRange.Bound;
+import org.apache.phoenix.query.QueryServices;
+import org.apache.phoenix.query.QueryServicesOptions;
 import org.apache.phoenix.schema.PColumn;
 import org.apache.phoenix.schema.RowKeySchema;
 import org.apache.phoenix.schema.SortOrder;
 import org.apache.phoenix.schema.TableRef;
 import org.apache.phoenix.schema.types.PDataType;
 import org.apache.phoenix.schema.types.PInteger;
+import org.apache.phoenix.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
 import org.apache.phoenix.util.MetaDataUtil;
 import org.apache.phoenix.util.ScanUtil;
 import org.apache.phoenix.util.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 public abstract class ExplainTable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExplainTable.class);
     private static final List<KeyRange> EVERYTHING = Collections.singletonList(KeyRange.EVERYTHING_RANGE);
     public static final String POINT_LOOKUP_ON_STRING = "POINT LOOKUP ON ";
+    public static final String REGION_LOCATIONS = " (region locations = ";
+
     protected final StatementContext context;
     protected final TableRef tableRef;
     protected final GroupBy groupBy;
     protected final OrderBy orderBy;
     protected final HintNode hint;
     protected final Integer limit;
     protected final Integer offset;
+    private final ExecutorService executorService;
```

Comment:
Why do we need an executor and async processing?

Reply:
this is because we want to have our own timeouts rather than relying on hbase client timeouts. 
if for some reason, we can't get region location due to some transient issue, client will keep retrying based on retry configs, but it would not be appropriate to block output of explain plan just because of that.